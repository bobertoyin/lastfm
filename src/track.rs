use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::{
    lfm_date::LfmDate,
    types::{Album, Artist},
};

#[derive(Deserialize, Serialize, Debug)]
pub struct Track {
    pub artist: Artist,
    pub album: Album,
    pub name: String,
    pub date: Option<LfmDate>,
    pub url: String,
    pub mbid: String,
    #[serde(deserialize_with = "de_now_playing", rename = "@attr", default)]
    pub now_playing: bool,
}

fn de_now_playing<'de, D>(deserializer: D) -> Result<bool, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let attrs: HashMap<String, String> = Deserialize::deserialize(deserializer)?;
    Ok(attrs.get("nowplaying").unwrap_or(&"false".to_string()) == "true")
}

// TODO: write test
