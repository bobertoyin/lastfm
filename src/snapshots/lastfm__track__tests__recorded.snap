---
source: src/track.rs
expression: track
---
Recorded(
    RecordedTrack {
        artist: Artist {
            image: ImageSet {
                small: Some(
                    "https://lastfm.freetls.fastly.net/i/u/34s/2a96cbd8b46e442fc41c2b86b821562f.png",
                ),
                medium: Some(
                    "https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png",
                ),
                large: Some(
                    "https://lastfm.freetls.fastly.net/i/u/174s/2a96cbd8b46e442fc41c2b86b821562f.png",
                ),
                extralarge: Some(
                    "https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png",
                ),
            },
            name: "Augustana",
            url: "https://www.last.fm/music/Augustana",
        },
        name: "Ash and Ember",
        image: ImageSet {
            small: Some(
                "https://lastfm.freetls.fastly.net/i/u/34s/65e46f0cb1864dc0cdb0c00db7ec8295.jpg",
            ),
            medium: Some(
                "https://lastfm.freetls.fastly.net/i/u/64s/65e46f0cb1864dc0cdb0c00db7ec8295.jpg",
            ),
            large: Some(
                "https://lastfm.freetls.fastly.net/i/u/174s/65e46f0cb1864dc0cdb0c00db7ec8295.jpg",
            ),
            extralarge: Some(
                "https://lastfm.freetls.fastly.net/i/u/300x300/65e46f0cb1864dc0cdb0c00db7ec8295.jpg",
            ),
        },
        album: "Life Imitating Life",
        url: "https://www.last.fm/music/Augustana/_/Ash+and+Ember",
        date: 2023-02-13T10:28:12Z,
    },
)
